python/cuml/dask/decomposition/base.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from cuml.dask.common import raise_exception_from_futures
+from cuml.dask.common.comms import worker_state, CommsContext
+
+from cuml.dask.common.input_utils import to_output
+
+from cuml.dask.common.part_utils import flatten_grouped_results
+
+from dask.distributed import wait
+
+from uuid import uuid1
+
+from cuml.dask.common.base import BaseEstimator
+from cuml.dask.common.input_utils import DistributedDataHandler
+
+
+class BaseDecompositionFitMixin(BaseEstimator):
+
+    def __init__(self, model_func, client=None, **kwargs):
+        """
+        Constructor for distributed PCA model
+        """
+        super(BaseDecompositionFitMixin, self).__init__(client, **kwargs)
+        self._model_func = model_func
+
+        # define attributes to make sure they
+        # are available even on untrained object
+        self.local_model = None
+        self.components_ = None
+        self.explained_variance_ = None
+        self.explained_variance_ratio_ = None
+        self.singular_values_ = None
+
+    @staticmethod
+    def _func_fit(m, dfs, M, N, partsToRanks, rank, transform):
+        return m.fit(dfs, M, N, partsToRanks, rank, transform)
+
+    def _fit(self, X, _transform=False):
+        """
+        Fit the model with X.
+
+        Parameters
+        ----------
+        X : dask cuDF input
+
+        """
+
+        X = self.client.persist(X)
+
+        wait(X)
+
+        data = DistributedDataHandler.single(data=X, client=self.client)
+        self.datatype = data.datatype
+
+        comms = CommsContext(comms_p2p=False)
+        comms.init(workers=data.workers)
+
+        data.calculate_parts_to_sizes(comms)
+
+        M = data.total_rows
```

Comment:
can we make these n_rows and n_cols? I always think of N as n_rows anyways so I find this a little confusing.